internal/gps/source_cache_bolt.go

```diff
@@ -0,0 +1,437 @@
+//	Key/Values: PackageOrErr fields
+//
+// c) Revision-versions buckets contain lists of version values:
+//
+//	Sub-Bucket: "versions:<timestamp>"
+//	Keys: "<sequence_number>"
+//	Values: "branch:<branch>", "defaultBranch:<branch>", "ver:<version>"
+type singleSourceCacheBolt struct {
+	ProjectRoot
+	db     *bolt.DB
+	epoch  int64       // getters will not return values older than this unix timestamp
+	logger *log.Logger // info logging
+}
+
+// newBoltCache returns a new singleSourceCacheBolt backed by a project's BoltDB file under the cache directory.
+func newBoltCache(cd string, pi ProjectIdentifier, epoch int64, logger *log.Logger) (*singleSourceCacheBolt, error) {
+	path := sourceCachePath(cd, pi.normalizedSource()) + ".db"
+	dir := filepath.Dir(path)
+	if fi, err := os.Stat(dir); os.IsNotExist(err) {
+		if err := os.MkdirAll(dir, os.ModeDir|os.ModePerm); err != nil {
+			return nil, errors.Wrapf(err, "failed to create source cache directory: %s", dir)
+		}
+	} else if err != nil {
+		return nil, errors.Wrapf(err, "failed to check source cache directory: ", dir)
+	} else if !fi.IsDir() {
+		return nil, errors.Wrapf(err, "source cache path is not directory: %s", dir)
+	}
+	db, err := bolt.Open(path, 0600, &bolt.Options{Timeout: 1 * time.Second})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to open BoltDB cache file %q", path)
+	}
+	return &singleSourceCacheBolt{
+		ProjectRoot: pi.ProjectRoot,
+		db:          db,
+		epoch:       epoch,
+		logger:      logger,
+	}, nil
+}
+
+// close releases all database resources.
+// Must not be called concurrently with any other methods.
+func (s *singleSourceCacheBolt) close() error {
+	return errors.Wrapf(s.db.Close(), "error closing Bolt database %q", s.db.String())
+}
+
+func (s *singleSourceCacheBolt) setManifestAndLock(rev Revision, ai ProjectAnalyzerInfo, m Manifest, l Lock) {
+	err := s.updateBucket("rev:"+string(rev), func(b *bolt.Bucket) error {
+		pre := "info:" + ai.String() + ":"
+		if err := cachePrefixDelete(b, pre); err != nil {
+			return err
+		}
+		info, err := b.CreateBucket(cacheTimestampedKey(pre, time.Now()))
+		if err != nil {
+			return err
+		}
+
+		// Manifest
+		mb, err := info.CreateBucket([]byte("manifest"))
+		if err != nil {
+			return err
+		}
+		if err := cachePutManifest(mb, m); err != nil {
+			return errors.Wrap(err, "failed to put manifest")
+		}
+		if l == nil {
+			return nil
+		}
+
+		// Lock
+		lb, err := info.CreateBucket([]byte("lock"))
+		if err != nil {
+			return err
+		}
+		return errors.Wrap(cachePutLock(lb, l), "failed to put lock")
+	})
+	if err != nil {
+		s.logger.Println(errors.Wrapf(err, "failed to cache manifest/lock for revision %q, analyzer: %v", rev, ai))
+	}
+}
+
+func (s *singleSourceCacheBolt) getManifestAndLock(rev Revision, ai ProjectAnalyzerInfo) (m Manifest, l Lock, ok bool) {
+	err := s.viewBucket("rev:"+string(rev), func(b *bolt.Bucket) error {
+		info := cacheFindLatestValid(b, "info:"+ai.String()+":", s.epoch)
```

Comment:
i haven't fully grokked how we're using these epoch markers, but given that revisions - and therefore the data we derive from them - are immutable, what's the reason to use an epoch marker here at all? it seems like this should basically be a permanent cache.

maybe there's a spot somewhere else where, for the M&L and ptrees, we use a value for the epoch that has the effect of making this basically a permanent cache, unless some magic value is passed in to clear the bucket, basically for maintenance purposes? or something?

Reply:
There was a point that I was not time-stamping manifests/locks or package trees, but then I convinced myself it was necessary again.  Now I can't think of any reason these wouldn't be immutable that wouldn't also warrant wiping the cache anyways (e.g. a new dep version produces a different package tree; or a `ProjectAnalyzerInfo` changes the way it parses the files, without changing its `Version`).

Happy to pull it back out. Simplifies a lot of code.

Reply:
> e.g. a new dep version

yep yep yep - i've been contemplating that gps needs its own internal versioning number that we include in these indices, and bump whenever we make a change that could have _any_ effect on how the gps interprets data on disk, eg in `ListPackages()`.

i believe what i settled on is what we see as the `solver-version` in `Gopkg.lock`, but that's a property of the `Solver`, not the `SourceManager`. maybe i made a mistake in not separating them out, but i think i probably did it that way because we effectively can't separate the two :/. still, not great. i'm going to have to think about this.

in any case, let's go with pulling the timestamp back out - i see this is changed, so perhaps you've already done as much :)